internal/auth/legacy.go

```diff
@@ -0,0 +1,96 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/platformsh/cli/internal/legacy"
+	"golang.org/x/oauth2"
+)
+
+type legacyCLITokenSource struct {
+	ctx     context.Context
+	cached  *oauth2.Token
+	wrapper *legacy.CLIWrapper
+	mu      sync.Mutex
+}
+
+func (ts *legacyCLITokenSource) getLegacyCLIToken() (*oauth2.Token, error) {
+	bt := bytes.NewBuffer(nil)
+	ts.wrapper.Stdout = bt
+	if err := ts.wrapper.Exec(ts.ctx, "auth:token", "-W"); err != nil {
+		return nil, fmt.Errorf("cannot retrieve token: %w", err)
+	}
+
+	at, err := unsafeParseJWT(bt.String())
+
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse token: %w", err)
+	}
+
+	return &oauth2.Token{
+		AccessToken: bt.String(),
+		TokenType:   "Bearer",
+		Expiry:      time.Unix(at.ExpiresAt, 0),
+	}, nil
+}
+
+func (ts *legacyCLITokenSource) refreshToken() error {
+	ts.cached = nil
```

Comment:
I think `ts.mu` is guarding `ts.cached`, and probably refreshes in general, so it would need to be locked here.


```suggestion
	ts.mu.Lock()
	defer ts.mu.Unlock()
	ts.cached = nil
```

Reply:
Fixed by adding `unsafe*` internal functions, so that we don't have deadlocks.